firedrake/functionspaceimpl.py

```diff
@@ -321,6 +321,14 @@ def __iter__(self):
         return iter(self.subfunctions)
 
     def __getitem__(self, i):
+        from firedrake.functionspace import MixedFunctionSpace
```

Comment:
I think we should cache the mixed subspaces like we do for the non-mixed subspaces. Otherwise we get a "different" space every time we call with the same indices.